documentation-website/Writerside/topics/DAO-Entity-definition.topic

```diff
@@ -6,45 +6,228 @@
+            <a href="%BASE_API_URL%/exposed-dao/org.jetbrains.exposed.dao/-immutable-entity-class/index.html"><code>ImmutableEntityClass</code></a>
+            and
+            <a href="%BASE_API_URL%/exposed-dao/org.jetbrains.exposed.dao/-immutable-cached-entity-class/index.html"><code>ImmutableCachedEntityClass</code></a>
+            .
+        </p>
+        <p>
+            The <code>ImmutableCachedEntityClass</code> uses an internal cache to store entity loading
+            states by the associated database. This ensures that entity updates are synchronized with this class as
+            the lock object.
+        </p>
+        <p>
+            These classes can be used as companion objects when defining entities that are
+            meant to be immutable.
+        </p>
+        <code-block lang="kotlin" src="exposed-dao/src/main/kotlin/org/example/entities/CityEntity.kt" />
+    </chapter>
+    <chapter id="class-method-overrides" title="Class method overrides">
+        <p>
+            You can use override methods in both the <code>Entity</code> class and the <code>EntityClass</code>
+            companion object to extend functionality or manage entity behavior.
+        </p>
+        <p>
+            For example, here’s how to override the <code>delete()</code> method in a <code>User</code> entity:
+        </p>
+        <code-block lang="kotlin" src="exposed-dao/src/main/kotlin/org/example/entities/UserEntityWithOverride.kt" />
+        <p>
+            In this example, a custom message is printed before the <code>delete()</code> function of the superclass
+            (<code>IntEntity</code>) completes the deletion.
+        </p>
+    </chapter>
+    <chapter title="Field transformations" id="field-transformations">
+        <p>
+            As databases typically store only basic types, such as integers and strings, it's not always convenient to keep
+            the same simplicity on Data Access Object (DAO) level.
+        </p>
+        <p>
+            For example, you might need to parse JSON from a <code>VARCHAR</code> column, or retrieve values from a cache
+            based on data from the database.
+            In such cases, the preferred approach is to use column transformations.
+        </p>
+        <chapter title="Example: Defining an Unsigned Integer field" id="defining-unsigned-int-field">
+            <p>
+                Suppose that you want to define an unsigned integer field on an entity, but Exposed doesn't have such
+                column type yet. You can achieve this by using the following implementation:
+            </p>
+            <code-block lang="kotlin"><![CDATA[
+        object TableWithUnsignedInteger : IntIdTable() {
+            val uint = integer("uint")
+        }
+        class EntityWithUInt : IntEntity() {
+            var uint: UInt by TableWithUnsignedInteger.uint.transform({ it.toInt() }, { it.toUInt() })
+
+            companion object : IntEntityClass<EntityWithUInt>()
+        }
```

Comment:
Code is missing a few pieces to compile:
```kt
class EntityWithUInt(id: EntityID<Int>) : IntEntity(id) {
    var uint: UInt by TableWithUnsignedInteger.uint.transform({ it.toInt() }, { it.toUInt() })

    companion object : IntEntityClass<EntityWithUInt>(TableWithUnsignedInteger)
}
```

Reply:
Fixed and moved to a separate entity file. This function seems to be deprecated though. Is there a better example we can show here?

Reply:
Yes, the old DAO `transform()` is currently deprecated but it will not be removed from the API entirely until some issues with caching are fully resolved.
If all goes well, this entire 'Fields transformation' section would become redundant and we could redirect users to the [DSL transform section](https://jetbrains.github.io/Exposed/data-types.html#column-transformation).
But if the issues prove problematic, then it would be good to keep this example up in the event we rollback deprecation (or implement some other DAO-level change).